src/java.desktop/unix/classes/sun/awt/screencast/TokenStorage.java

```diff
@@ -0,0 +1,416 @@
+
+/**
+ * Helper class for persistent storage of ScreenCast restore tokens
+ * and associated screen boundaries.
+ *
+ * The restore token allows the ScreenCast session to be restored
+ * with previously granted screen access permissions.
+ */
+final class TokenStorage {
+
+    private TokenStorage() {}
+
+    private static final String REL_NAME =
+            ".awt/robot/screencast-tokens.properties";
+
+    private static final Properties PROPS = new Properties();
+    private static final Path PROPS_PATH;
+    private static final Path PROP_FILENAME;
+
+    static {
+        PROPS_PATH = setupPath();
+        if (PROPS_PATH != null) {
+            PROP_FILENAME = PROPS_PATH.getFileName();
+            if (SCREENCAST_DEBUG) {
+                System.out.println("Token storage: using " + PROPS_PATH);
+            }
+            setupWatch();
+        } else {
+            // We can still work with tokens,
+            // but they are not saved between sessions.
+            PROP_FILENAME = null;
+        }
+    }
+
+    private static Path setupPath() {
+        String userHome = System.getProperty("user.home", null);
+        if (userHome == null) {
+            return null;
+        }
+
+        Path path = Path.of(userHome, REL_NAME);
+        Path workdir = path.getParent();
+
+        if (!Files.exists(workdir)) {
+            try {
+                Files.createDirectories(workdir);
+            } catch (Exception e) {
+                if (SCREENCAST_DEBUG) {
+                    System.err.printf("Token storage: cannot create" +
+                                    " directory %s %s\n", workdir, e);
+                }
+                return null;
+            }
+        }
+
+        if (!Files.isWritable(workdir)) {
+            if (SCREENCAST_DEBUG) {
+                System.err.printf("Token storage: %s is not writable\n", workdir);
+            }
+            return null;
+        }
+
+        try {
+            Files.setPosixFilePermissions(workdir,
+                    Set.of(
+                            PosixFilePermission.OWNER_READ,
+                            PosixFilePermission.OWNER_WRITE,
+                            PosixFilePermission.OWNER_EXECUTE
+                    ));
+        } catch (IOException e) {
+            if (SCREENCAST_DEBUG) {
+                System.err.printf("Token storage: cannot set permissions " +
+                        "for directory %s %s\n", workdir, e);
+            }
+        }
+
+        if (Files.exists(path)) {
+            if (!setFilePermission(path)) {
+                return null;
+            }
+
+            readTokens(path);
+        }
+
+        return path;
+    }
+
+    private static boolean setFilePermission(Path path) {
+        try {
+            Files.setPosixFilePermissions(
+                    path,
+                    Set.of(PosixFilePermission.OWNER_READ,
+                            PosixFilePermission.OWNER_WRITE)
```

Comment:
alignment issue..Posix..OWNER_WRITE should be placed below Posix..OENER_READ

Also, why we aren't adding OWNER_EXECUTE here too like done above?

Reply:
Because execute is required on directories but not files unless they are programs.

Reply:
Exactly, without an execute bit on the directory we cannot access/add/remove its files.

Our file is not a script or executable that needs to be run.